python/lsst/meas/astrom/matcher_probabilistic.py

```diff
@@ -0,0 +1,437 @@
+        itemtype=str,
+        dictCheck=lambda x: len(x) == 2,
+        doc='Dict mapping sky coordinate columns to be converted to tract pixel columns',
+    )
+    mag_brightest_ref = pexConfig.Field(
+        dtype=float,
+        default=-np.inf,
+        doc='Bright magnitude cutoff for selecting reference sources to match.'
+            ' Ignored if column_order is None.'
+    )
+    mag_faintest_ref = pexConfig.Field(
+        dtype=float,
+        default=np.Inf,
+        doc='Faint magnitude cutoff for selecting reference sources to match.'
+            ' Ignored if column_order is None.'
+    )
+    mag_zeropoint_ref = pexConfig.Field(
+        dtype=float,
+        default=31.4,
+        doc='Magnitude zeropoint for computing reference source magnitudes for selection.'
+            ' Ignored if column_order is None.'
+    )
+    match_dist_max = pexConfig.Field(
+        dtype=float,
+        default=0.5,
+        doc='Maximum match distance. Units must be arcseconds if coords_spherical, '
+            'or else match those of column_*_coord[12] multiplied by coords_*_factor.',
+    )
+    match_n_max = pexConfig.Field(
+        dtype=int,
+        default=10,
+        optional=True,
+        doc='Maximum number of spatial matches to consider (in ascending distance order).',
+    )
+    match_n_finite_min = pexConfig.Field(
+        dtype=int,
+        default=3,
+        optional=True,
+        doc='Minimum number of pexConfig.Fields with a finite value to measure match likelihood',
+    )
+
+    order_ascending = pexConfig.Field(
+        dtype=bool,
+        default=False,
+        optional=True,
+        doc='Whether to order reference match candidates in ascending order of column_order '
+            '(should be False if the column is a flux and True if it is a magnitude.',
+    )
+
+
+@dataclass
+class CatalogExtras:
+    """Store frequently-reference (meta)data revelant for matching a catalog.
+
+    Parameters
+    ----------
+    catalog : `pandas.DataFrame`
+        A pandas catalog to store extra information for.
+    select : `numpy.array`
+        A numpy boolean array of the same length as catalog to be used for
+        target selection.
+    """
+    n: int
+    indices: np.array
+    select: np.array
+
+    coordinate_factor: float = None
+
+    def __init__(self, catalog: pd.DataFrame, select: np.array = None, coordinate_factor: float = None):
+        self.n = len(catalog)
+        self.select = select
+        self.indices = np.flatnonzero(select) if select is not None else None
+        self.coordinate_factor = coordinate_factor
+
+
+class MatcherProbabilistic:
+    """A probabilistic, greedy catalog matcher.
+
+    Parameters
+    ----------
+    config: `MatchProbabilisticConfig`
+        A configuration instance.
+    """
+    config: MatchProbabilisticConfig
+
+    def __init__(
+            self,
+            config: MatchProbabilisticConfig,
+    ):
+        self.config = config
+
+    def match(
```

Comment:
Code could use some inline comments describing what is happening at a give step.

Reply:
Also, would it make sense to break some of this function up into sub sections? That would certainly make the testing easier.

Reply:
I don't think there would be much benefit in breaking it up, as any of the sub-functions I could think of defining would have large numbers of arguments/return values but still not really make sense to call in any other context.